tests/ga/extension_actor.py

```diff
@@ -0,0 +1,465 @@
+    protocol._actors.extend(new_actors)
+
+    for actor in new_actors:
+        patched_popen = actor.patch_popen() # Like in get_protocol_and_handler, the patch_popen() and start() calls must
+        patched_popen.start()               # be interleaved.
+
+        # Keep track of the patches so the finally clause within get_protocol_and_handler can stop them.
+        protocol._patched_popens.append(patched_popen)
+    
+    protocol.set_http_handlers(http_get_handler=_generate_mock_http_get(protocol._actors[1:]),
+        http_put_handler=_generate_mock_http_put(protocol._actors))
+
+    distinct_names = set(actor.extension_info.name for actor in protocol._actors)
+    for name in distinct_names:
+        
+        max_ver = max(map(lambda actor: FlexibleVersion(actor.extension_info.version),
+            filter(lambda actor: actor.extension_info.name == name, protocol._actors)))
+
+        protocol.mock_wire_data.set_specific_extension_config_version(name, max_ver)
+    
+    protocol.mock_wire_data.set_incarnation(incarnation)
+    protocol.client.update_goal_state()
+
+
+class ActionSet(object):
+    """
+    A wrapper class for the possible actions that an extension much support.
+    """
+
+    def __init__(self, installAction, uninstallAction, updateAction, enableAction, disableAction, name, version):
+
+        # The keys below will be used to configure attributes for a mock; periods in such attributes are treated
+        # as attribute trees, which would break introspection.
+        formattedId = Formats.FormatExtensionDir(name, version).replace(".", "_")
+
+        self.delegate ={
+            "installCommand": dict(key="{0}_install".format(formattedId), action=installAction),
+            "uninstallCommand": dict(key="{0}_uninstall".format(formattedId), action=uninstallAction),
+            "updateCommand": dict(key="{0}_update".format(formattedId), action=updateAction),
+            "enableCommand": dict(key="{0}_enable".format(formattedId), action=Actions._wrap_with_write_status(enableAction, name, version)),
+            "disableCommand": dict(key="{0}_disable".format(formattedId), action=disableAction),
+        }
+
+    def items(self):
+        """
+        Keys are the various command names an extension must support, and values have the property layout { key, action }.
+
+        A wrapper around dict.items().
+        """
+        for key, action in self.delegate.items():
+            yield key, action
+    
+    def keys(self):
+        """
+        Yields the list of all possible command names an extension must support.
+
+        A wrapper around dict.keys().
+        """
+        for key in self.delegate.keys():
+            yield key
+    
+    def getKeyForCommand(self, cmd):
+        """
+        Returns the unique ID associated with a specific command name in this ActionSet.
+        cmd may be "installCommand", "enableCommand", etc.
+        """
+        return self.delegate.get(cmd, {}).get("key", None)
```

Comment:
is None an appropriate return value, or should this raise? seems like if somebody passes a command that does not exist that should be an error

Reply:
I think it might as well raise; nothing checks the returned value against None before attempting to access an attribute.